src/jvm/main/org/jetbrains/kotlinx/lincheck/ObjectTraverser.kt

```diff
@@ -39,107 +35,86 @@ import kotlin.coroutines.Continuation
  */
 internal fun enumerateObjects(obj: Any): Map<Any, Int> {
     val objectNumberMap = hashMapOf<Any, Int>()
-    enumerateObjects(obj, Collections.newSetFromMap(IdentityHashMap()), objectNumberMap)
-
+    enumerateObjects(obj, objectNumberMap)
     return objectNumberMap
 }
 
 /**
  * Recursively traverses an object to enumerate it and all nested objects.
  *
  * @param obj object to traverse
- * @param processedObjects a set of already processed objects. Required in case of cyclic references.
  * @param objectNumberMap result enumeration map
  */
-private fun enumerateObjects(obj: Any, processedObjects: MutableSet<Any>, objectNumberMap: MutableMap<Any, Int>) {
+private fun enumerateObjects(obj: Any, objectNumberMap: MutableMap<Any, Int>) {
     if (obj is Class<*> || obj is ClassLoader) return
-    if (!processedObjects.add(obj)) return
-    val objectNumber = getObjectNumber(obj.javaClass, obj)
-    objectNumberMap[obj] = objectNumber
-
-    var clazz: Class<*>? = obj.javaClass
-    if (clazz!!.isArray) {
-        if (obj is Array<*>) {
-            obj.forEach { element -> element?.let { enumerateObjects(it, processedObjects, objectNumberMap) } }
+    objectNumberMap[obj] = getObjectNumber(obj.javaClass, obj)
+
+    val processObject: (Any?) -> Any? = processObject@{ value: Any? ->
+        if (value == null || value is Class<*> || value is ClassLoader) return@processObject null
```

Comment:
Is it possible to organize the code flow without labels? The current version is hard to read.